### ISSUE TITLE
[unstable option] unstable option: wrap_comments

### ISSUE THREAD

**Issue by scampi** [rust-lang/rustfmt#3347]
Tracking issue for unstable option: wrap_comments

**Comment by carllerche** [comment]
What is left to stabilize this?

**Comment by scampi** [comment]
There is now a documented process describing stabilisation so we'd have to go through the conditions set there.

**Comment by cavedweller** [comment]
Is there a link to the stabilization process? I can't find it from the README.

I'd like to see us go through this process for the `wrap_comments` feature

**Comment by calebcartwright** [comment]
Not sure if it's up to date, but the process for stabilizing options is defined in https://github.com/rust-lang/rustfmt/blob/master/Processes.md

@scampi & @topecongiro - correct me if I'm wrong but my understanding is that the current focus is on rustfmt 2.0, and that stabilization of some of these currently-unstable config options won't come til after the 2.0 release

**Comment by jhpratt** [comment]
FWIW while I'd like to see this stabilized, this option can also break markdown. Last time I tried, a large markdown table I had was wrapped, which meant it was no longer valid markdown.

**Comment by bitdivine** [comment]
Is there a way of saying "Do not wrap this comment"?  We have a similar problem with long maths formulae.  It's not a blocker, it just takes a bit of effort to fix up the formulae afterwards, however it would be nice if we could make exceptions for some blocks of text.  I suppose the markdown could be checked programatically before and after each conversion and the formatting could be skipped if the markdown is broken.  That won't catch the formulae but it might help in your case @jhpratt .  Maybe there is a nice markdown formatter out there in the wild that understands when it is looking at a table.

**Comment by daniel5151** [comment]
Any updates on this?

**Comment by jhpratt** [comment]
@daniel5151 Updates are posted as they come in. What you see in this thread is the current status.

**Comment by Nemo157** [comment]
Would it be plausible to split this into `wrap_comments` and `wrap_doc_comments`, and make `wrap_doc_comments` markdown aware?

**Comment by calebcartwright** [comment]
> Would it be plausible to split this into wrap_comments and wrap_doc_comments, and make wrap_doc_comments markdown aware?

Could you elaborate a bit on the motivation? Is it just about wanting `wrap_comments` stabilized more quickly or are there cases where you'd want to wrap doc comments but not other comments or vice versa?

**Comment by Nemo157** [comment]
I actually don't care about normal comments much, I want to be able to have rustfmt wrap my doc comments. My thought was that normal comments aren't required to be markdown, so it would make sense to split the implementation.

**Comment by jhpratt** [comment]
Just filed an issue regarding the breakage of markdown tables. It looks like this was fixed at some point in the past, then we regressed.

**Comment by calebcartwright** [comment]
> It looks like this was fixed at some point in the past, then we regressed.

That is incorrect. There are different branches in this repository with what at this point are fairly divergent versions of rustfmt. A fix was made to the master branch, but wasn't applied (much less released) on any branches associated with current/distributed versions of rustfmt.

Going to hide these last two comments to avoid any confusion

**Comment by obsgolem** [comment]
The current implementation of this feature doesn't reflow short comments back into the lines above. Is this something that could be implemented?

**Comment by dcow** [comment]
This feature does not remove a leading blank comment line but it removes trailing blank lines. 

```
///
/// Foo is a newtype for bar that can be passed to baz(foo) ...end of long line.
///
struct Foo ...
```
becomes
```
///
/// Foo is a newtype for bar that can be passed to baz(foo)
/// ...end of long line.
struct Foo ...
```

**Comment by tgross35** [comment]
Just brainstorming here... there's an actively maintained library [`comrak`](https://crates.io/crates/comrak) that does markdown formatting, and allows setting line width. Could it make sense to use this on doc comments? It already has to be aware of things like tables and other markdown constructs, so I feel like it would do a better job out of the box than the current implementation.

Basically it seems like rustdoc needs some markdown-aware comment formatting, and it doesn't make sense to reinvent the wheel

**Comment by luxalpa** [comment]
The reason I would prefer wrap_doc_comments is because sometimes I comment out my code temporarily, and then the normal `wrap_comments` can do some very unfortunate wrapping (particularly on macros).

**Comment by VorpalBlade** [comment]
It would be good to be able to turn this off for specific comments or files. My use case for this is things like `clap`, `bpaf` or `strum` where doc strings are used to generate command line help output and I want precise control over the formatting *in just those comments*, while elsewhere I don't want to have to think about this and just have rustfmt do a sensible default.

**Comment by sww1235** [comment]
@VorpalBlade would using `#[rustfmt::skip]` work for your use case?

**Comment by c-git** [comment]
> I would prefer a `wrap_doc_comments` because sometimes I comment out my code temporarily, and then the normal `wrap_comments` can do some very unfortunate wrapping (particularly on macros).

+1 for this. I just came here this. I've been using wrap_comments daily now and it's been really good. I've been happy just typing and all the wrapping gets taken care off automatically until today when I had long lines of code with calls to the error tracing macro that I commented out temporarily and now tried to uncomment to find that all my macros got line wrapped. I have no idea how this could be addressed but this is IMO a surprising problem. Obvious after you think about it but not acceptable on stable rust IMO. **New users** coming in would do as they normally do comment out code and be surprised when they try to uncomment it.

**Comment by JarredAllen** [comment]
I have a proposal for an alternate solution to `wrap_comments` wrapping people's code:

Personally, whenever I have a comment that I want to persist in the code, I use `// ..`-style comments. But when I'm commenting out some snippet of code, I instead use `/* .. */`-style comments. So, an option to format only the `// ..`-style comments would (for me, at least) fix the issue of it rewrapping macros while temporarily commented-out.

I'm not sure how many people use the same scheme as me, maybe y'all should 👍 if this option would work for y'all, and 👎 if you use comments differently such that it wouldn't?

**Comment by Lokathor** [comment]
that works and i used to to it, but some part of rust screams at unknown configs by default these days.

**Comment by c-git** [comment]
Those are great work a rounds but I'm saying that for people new to rust that is not exactly approachable. I'm not the only person that highlights some lines of code in my IDE then presses <kbd>Ctrl</kbd> + <kbd>/</kbd> to comment them out. It's a surprising thing to have that then get broken by the formatter so that it cannot just be uncommented

**Comment by ia0** [comment]
Is it possible to add #6180 as a sub-issue?

**Comment by honestlysamuk** [comment]
I found a bug using this feature flag where it involved comment blocks. There were occasions where I had something like

```
// a really long comment that goes over my comment_width value immediately followed by a code block
// ```
// some code
// ```

```

which then transformed to

```
// a really long comment that goes over my comment_width value
// immediately followed by a code block```
// some code
// ```
```

I think that this is the markdown aware part of the doc comment above. Clippy then lints our docs and warns on this.
